### PR TITLE
Mapillary street signs

### PR DESCRIPTION
**Pull request by bhousel**
Cherry picked the important commits from #2703..

http://blog.mapillary.com/update/2015/06/02/id-signs.html

### CODE REVIEW

File: js/id/renderer/mapillary_sign_layer.js [hunk 1]
```diff
@@ -0,0 +1,177 @@
+iD.MapillarySignLayer = function(context) {
+    var roundedProjection = iD.svg.RoundProjection(context.projection),
+        urlSearch = 'https://a.mapillary.com/v2/search/im/geojson/or',
+        urlImage = 'http://mapillary.com/map/im/',
+        urlThumb = 'https://d1cuyjsrcm0gby.cloudfront.net/',
+        clientId = 'NzNRM2otQkR2SHJzaXJmNmdQWVQ0dzoxNjQ3MDY4ZTUxY2QzNGI2',
```
Comment (peterneubauer) [inline js/id/renderer/mapillary_sign_layer.js]: please change to 
    clientId='NzNRM2otQkR2SHJzaXJmNmdQWVQ0dzo1ZWYyMmYwNjdmNDdlNmVi'

File: js/id/renderer/mapillary_image_layer.js [hunk 2]
```diff
@@ -1,5 +1,10 @@
-iD.MapillaryLayer = function (context) {
-    var enable = false,
+iD.MapillaryImageLayer = function (context) {
+    var roundedProjection = iD.svg.RoundProjection(context.projection),
+        urlSearch = 'https://a.mapillary.com/v2/search/s/geojson',
+        urlImage = 'http://mapillary.com/map/im/',
+        urlThumb = 'https://d1cuyjsrcm0gby.cloudfront.net/',
+        clientId = 'NzNRM2otQkR2SHJzaXJmNmdQWVQ0dzoxNjQ3MDY4ZTUxY2QzNGI2',
```
Comment (peterneubauer) [inline js/id/renderer/mapillary_image_layer.js]: please change to 
    clientId='NzNRM2otQkR2SHJzaXJmNmdQWVQ0dzo1ZWYyMmYwNjdmNDdlNmVi'

**Comment by ireun** [comment]
Hello, will this ever be applied to ID?

**Comment by bhousel** [comment]
> Hello, will this ever be applied to ID?

Yes, it will be applied.  I got pretty far along with the cleanup as of July, then personal things got in the way and I forgot about this..  Thanks for being patient.

**Comment by ireun** [comment]
No problem, I'm waiting for this :)

**Comment by peterneubauer** [comment]
Awesome, thanks!

**Comment by bhousel** [comment]
@peterneubauer I have this updated to run on the latest version of iD, but it looks like the object recognition api has moved a little bit.

Using this endpoint:  `https://a.mapillary.com/v2/search/im/geojson/or'`

I am getting back features like:

```json
{
  "type": "Feature",
  "geometry": {
    "type": "Point",
    "coordinates": [-74.45660313,40.65164319]
  },
  "properties": {
    "key": "xxxxx",
    "country": "United States",
    "boundary": {
      "min_lat": 40.65164319,
      "max_lat": 40.65164319,
      "min_lon": -74.45660313,
      "max_lon": -74.45660313
    },
    "rects": [
      {
        "package": "trafficsign_us_3.0",
        "score": "1.07248174363",
        "rect": [
          0.65234375,
          0.44921875,
          0.6796875,
          0.4856770833333333
        ],
        "type": "warning--traffic-merges-right-ahead--us"
      }
    ]
  }
}
```

That `type` field:  `warning--traffic-merges-right-ahead--us`
previously was a key that matched up exactly with the object in `traffico-release-0.1.5\global-patched.json`
and looks like that is not the case anymore.

Is that `"package": "trafficsign_us_3.0"` out of sync with `traffico`?

**Comment by peterneubauer** [comment]
Will check it out tomorrow and update this pr of necessary...

**Comment by bhousel** [comment]
> Will check it out tomorrow and update this pr of necessary...

great thanks!
Around here is the code that used to work and now doesn't:

https://github.com/openstreetmap/iD/blob/3c5ffec393e11be345eae4d022e2f5f9dd41df2e/js/id/renderer/mapillary_sign_layer.js#L96-L117

If I replace that inner html code with something that matches up with traffico, I get signs rendered again:

```javascript
.html(function(d) {
   var sign_html = signs_defs['warning_hairpin_curve']; //signs_defs[d.signs[0].type];
   return sign_html;
});
```

screenshot 2015-11-09 16 26 26

**Comment by peterneubauer** [comment]
So,
the mapping of the signs has been changing in order to cover all countries with consistent naming conventions, including some backwards compatibility. Let me get some code to read the HTML base off the detection-package and the detected sign, so this gets more stable. will take some day though...

**Comment by peterneubauer** [comment]
Created a new PR with a first update to check, https://github.com/openstreetmap/iD/pull/2828

**Comment by riordan** [comment]
Just a :+1: on getting this into iD. Having easy access to adding signs would be amazing.

**Comment by peterneubauer** [comment]
Yes, is there anything I can do to help merge this?

**Comment by bhousel** [comment]
> Yes, is there anything I can do to help merge this?

Not yet, but I will let you know, I'm sorry it keeps getting pushed back.. 

I just released 1.8.4 which contains a lot of little fixes for usability and stability, so lets shoot for late  Jan/early Feb for a next release that includes this Mapillary street signs PR.

**Comment by peterneubauer** [comment]
So, I have another update coming in - using Mapbox vector tiles and the all-new Mapillary viewer in iD, see http://mapillary.github.io/mapillary-js/ .

Where should I start from in order not to destroy merged commits? @bhousel ?

**Comment by bhousel** [comment]
> Where should I start from in order not to destroy merged commits? @bhousel ?

I'll pick this up again this weekend and let you know Monday.

**Comment by peterneubauer** [comment]
Ok! I have a local branch with the changes, but need some short help with
catching the events on the the vector tiles added to d3, and zoom
projection . Let me know when you are done so I can apply these for your
feedback.

**Comment by peterneubauer** [comment]
Any progress? I'm happy to start wherever you have a baseline ...

**Comment by bhousel** [comment]
Ok, we're getting closer... we have street signs again! 

There are some things about this that are still pretty hacky, so I'll do some cleanup tomorrow and expect to get it merged to master this week.  @peterneubauer can you wait until this is merged to send your request to upgrade the image viewer?

screenshot 2016-02-07 23 39 21

**Comment by peterneubauer** [comment]
Yes, happy to wait, commit a fork to master and get your help later this week. Thanks @bhousel for picking this up!

**Comment by bhousel** [comment]
Just a quick update on this.  I've been able to move things forward a lot this week and it's looking pretty awesome:

mapillary_streetsigns


Here are some cool things that the Mapillary data layers do now:
* Common Mapillary code moved to service layer
* Fetch data asynchronously across visible tiles
* Cache results in rbush
* Trigger redraws as data becomes available

These are big changes from the existing code, but this work can pave the way for us to implement OSM notes, or QA layers, or any other data overlay using exactly the same tricks.

I ran into some cross browser issues with how the Mapillary traffic icon font is implemented using the SVG `foreignObject` element.  This is not an easy element to make work across all browsers, but I did get it working mostly:
* Chrome 48.0.2564.103 - works great
* Firefox 44.0.1 - works great
* Safari Nightly - `foreignObject` icons not positioned correctly due to [this Webkit bug](https://bugs.webkit.org/show_bug.cgi?id=71819), [more info](http://stackoverflow.com/questions/8185845/svg-foreignobject-behaves-as-though-absolutely-positioned-in-webkit-browsers)
* IE11 - does not work - no `foreignObject` support at all
* Edge - works great

So I'm going to remove the Mapillary street signs option if the user is using an unsupported browser.

**Comment by peterneubauer** [comment]
Awesome! Do you want to change the image viewer to the new MapillaryJS?
And, do you think we should use vector tiles instead of markers to get the
number of svg layers down for imagery?

**Comment by pnorman** [comment]
Do you want me to check it on the torturous I-5 regions where the Mapillary API returns thousands of pictures over hundreds of km for even very small bounding boxes?

**Comment by peterneubauer** [comment]
Well,
I think we should go Vector Maps, the markers are not going to scale. I
have that code sort of working but need help with the zoom/pan functions on
the vector tiles. I can add it for your feedback when I get a merged offset
for the current state.

**Comment by bhousel** [comment]
> Do you want me to check it on the torturous I-5 regions where the Mapillary API returns thousands of pictures over hundreds of km for even very small bounding boxes?

Yes, I'd like to test this.. I've been testing over Manhattan which is pretty dense also.  What has helped tremendously has been chopping the view into tiles and loading images within those tile bounds - rather than before we were just loading all sequences that happen to pass through the viewport (and refetching them with every zoom/pan).  

Fetching many thousands of images is not a problem - rbush is really efficient at indexing them.  But zooming out to z12 basically turns this into a browser svg rendering stress test.  

Some things I'm going to try:
* we should probably set a minzoom for these layers.. maybe z11ish?
* pad each image with some space and collisionbox cull them, the same way we do with labels

This allows people to see where data is, and they get more of it as they zoom in.

**Comment by bhousel** [comment]
For the curious, here is what 36530 images and 1732 street signs look like at z11.  
Chrome handles this _surprisingly_ well.  

I'm getting about 5fps, but the fans aren't making any noise.  It helps a lot that all the OSM data is not visible at this zoom. 

screenshot 2016-02-11 10 17 40

**Comment by peterneubauer** [comment]
Sounds good. In that case, just adding the new viewer to that will be
trivial.

**Comment by riordan** [comment]
I see you've encountered New York City's ~1 million street signs.

**Comment by slhh** [comment]
Malmö (Sweden), where Mappillary is based, would be a good area for testing with high density.
Jena (Germany) is also covered quite well.

Btw, It wouldn't make sense to display a subset of the available images only without using spacial filtering. Otherwise areas near a frequently used motorway look like they were not covered by Mappilary even if they are.

**Comment by peterneubauer** [comment]
Or,
use the vector tiles. They allow client side filtering etc without hitting
the server, and are not creating a lot of load on the client. See e.g.
http://bl.ocks.org/knikel/010115b08ebe30baba86

**Comment by bhousel** [comment]
> Btw, It wouldn't make sense to display a subset of the available images only without using spacial filtering. Otherwise areas near a frequently used motorway look like they were not covered by Mappilary even if they are.

Yes, this is the issue I'm going to solve today..

> Or, use the vector tiles. They allow client side filtering etc without hitting
> the server, and are not creating a lot of load on the client.

Yes, we'll get there, but not with this PR..

**Comment by bhousel** [comment]
Excited to finally merge this!

**Comment by peterneubauer** [comment]
Yay!

**Pull request merged**